### PR TITLE
[FLINK-5364] Rework JAAS configuration to support user-supplied entries

### PR DESCRIPTION
**Pull request by EronWright**
Fixes FLINK-5364, FLINK-5361, FLINK-5350, FLINK-5055

CC @tillrohrmann

### CODE REVIEW

**Comment by StephanEwen** [comment]
@EronWright thanks a lot for this fix, I am looking though it now.

From the description in the JIRAs, I take that this adds the logic that reads custom JAAS security configurations and uses Flink's internal one (Hadoop UGI based) for fallback default?

This should be relevant for the `master` and the 1.2 release branch, correct?

File: flink-runtime/src/main/java/org/apache/flink/runtime/security/modules/JaasModule.java [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+		final File jaasConfFile;
+		try {
+			Path jaasConfPath = Files.createTempFile("jaas-", ".conf");
+			try (InputStream resourceStream = JaasModule.class.getClassLoader().getResourceAsStream(JAAS_CONF_RESOURCE_NAME)) {
+				Files.copy(resourceStream, jaasConfPath, StandardCopyOption.REPLACE_EXISTING);
+			}
+			jaasConfFile = jaasConfPath.toFile();
+			jaasConfFile.deleteOnExit();
+		} catch (IOException e) {
+			throw new RuntimeException("unable to generate a JAAS configuration file", e);
```
Comment (StephanEwen) [inline flink-runtime/src/main/java/org/apache/flink/runtime/security/modules/JaasModule.java]: See above - why not declare that this method may throw an `IOException`?

**Comment by EronWright** [comment]
@StephanEwen yes, this PR does the following:
1. isolate the configuration of kerberos credentials from how those credentials are used
2. modularize the security code to reflect the independent aspects of configuring Hadoop vs. JAAS based on the credential.
3. make explicit as to which JAAS login contexts are provided the cluster's credential
4. incorporate the user-supplied JAAS (to satisfy the MapR scenario)
5. update `config.md` and `internals/security.md`

The main user impact is that one must explicitly share the credential with specific contexts - the 'KafkaClient' login context, ZooKeeper 'Client' context, etc, based on whether the corresponding service is kerberized.  Earlier we had tried to always share the credential to all JAAS contexts, but this caused problems.     

With this patch, I believe the design goal is met of allowing Kerberos to be independently enabled for any external connection.  For example, a Kerberized Kafka + a non-Kerberized Hadoop is a valid scenario.

**Comment by StephanEwen** [comment]
@EronWright Thanks for explaining. It figured it out concurrently by looking at the updated documentation you wrote ;-) The docs are good, helped a lot!

**Comment by EronWright** [comment]
Thanks for the review, I'll address these comments ASAP.

File: flink-runtime/src/main/java/org/apache/flink/runtime/security/SecurityUtils.java [hunk 2]
```diff
@@ -71,163 +64,93 @@
...
+					module.uninstall();
 				}
-			}
-
-			if (!loginUser.hasKerberosCredentials()) {
-				//throw an error in non-yarn deployment if kerberos cache is not available
-				if (!delegationToken) {
-					LOG.error("Hadoop Security is enabled but current login user does not have Kerberos Credentials");
-					throw new RuntimeException("Hadoop Security is enabled but current login user does not have Kerberos Credentials");
+				catch(UnsupportedOperationException e) {
```
Comment (EronWright) [inline flink-runtime/src/main/java/org/apache/flink/runtime/security/SecurityUtils.java]: The uninstall isn't used in production, only in test code.   Throwing gives more information to the unit test.  Hopefully getting rid of the warning will suffice for now.

File: flink-runtime/src/main/java/org/apache/flink/runtime/security/modules/JaasModule.java (hunk 1 above)
Comment (EronWright) [inline flink-runtime/src/main/java/org/apache/flink/runtime/security/modules/JaasModule.java]: In this case, I do consider a failure to generate the default config a programming error; it simply extracts an in-built resource as a temp file.   I definitely agree with your general point.

**Comment by EronWright** [comment]
@StephanEwen updated based on feedback, thanks again.

**Pull request closed without merging**